sycl/doc/design/PrivateAlloca.md

```diff
@@ -0,0 +1,191 @@
+
+`private_alloca` and `aligned_private_alloca` calls are defined as builtin
+aliases of `__builtin_intel_sycl_alloca` and
+`__builtin_intel_sycl_alloca_with_align` respectively using the
+`clang::builtin_alias` attribute. This way, calls to these functions are handled
+as calls to the builtins in the frontend. These builtins cannot be called
+directly, only via their aliases defined in the SYCL headers.
+
+```c++
+template <typename ElementType, auto &SizeSpecName,
+          access::decorated DecorateAddress>
+__SYCL_BUILTIN_ALIAS(__builtin_intel_sycl_alloca)
+[[__sycl_detail__::__uses_aspects__(aspect::ext_oneapi_private_alloca)]] private_ptr<
+    ElementType, DecorateAddress> private_alloca(kernel_handler &kh);
+
+template <typename ElementType, std::size_t Alignment,
+          auto &SizeSpecName, access::decorated DecorateAddress>
+__SYCL_BUILTIN_ALIAS(__builtin_intel_sycl_alloca_with_align)
+[[__sycl_detail__::__uses_aspects__(aspect::ext_oneapi_private_alloca)]] private_ptr<
+    ElementType, DecorateAddress> aligned_private_alloca(kernel_handler &kh);
+```
+
+
+As builtins cannot be passed template arguments, the original function
+declaration is used both in Sema and CodeGen. See the original source for
+documentation of each of these template parameters.
+
+Following `__builtin_alloca_with_align`, `Alignment` must be lower than
+`std::numeric_limits<int32_t>::max() / 8`.
+
+### `llvm.sycl.alloca.*` intrinsic
+
+The builtins mentioned in the previous section are represented via the new
+`llvm.sycl.alloca` intrinsic in code generation. This intrinsic receives three
+arguments encoding the specialization constant used as array length,
+corresponding to the arguments received by builtins implementing SYCL 2020
+specialization constants; a type hint argument encoding the allocation element
+type, and the required alignment, which must be an immediate argument.
+
+```llvm
+declare ptr @llvm.sycl.alloca.p0.p4.p4.p4.f32(ptr addrspace(4), ptr addrspace(4), ptr addrspace(4), float, i64)
+declare ptr @llvm.sycl.alloca.p0.p4.p4.p4.f64(ptr addrspace(4), ptr addrspace(4), ptr addrspace(4), double, i64)
+```
+
+The alignment argument is set to `alignof(ElementType)` for `private_alloca`;
+`Alignment` is used for `aligned_private_alloca`.
+
+The `private_alloca` call in the use-case above is compiled to the following
+LLVM IR:
+
+```llvm
+@_ZL4size = internal addrspace(1) constant %"class.sycl::_V1::specialization_id" { i64 1 }, align 8
+@__usid_str.1 = private unnamed_addr constant [31 x i8] c"uid2c9b8e1a387f5dba____ZL4size\00", align 1
+...
+%alloca.i = tail call ptr @llvm.sycl.alloca.p0.p4.p4.p4.f32(ptr addrspace(4) addrspacecast (ptr @__usid_str.1 to ptr addrspace(4)), ptr addrspace(4) addrspacecast (ptr addrspace(1) @_ZL4size to ptr addrspace(4)), ptr addrspace(4) null, float 0.000000e+00, i64 4)
+%3 = addrspacecast ptr %alloca.i to ptr addrspace(4)
+```
+
+**Note:** the third argument is set to `null` for now. That argument will be set
```

Comment:
If you already have a design outline for emulation path, it would be good to see it here as well, no need to wait for the implementation to be complete

Reply:
This design depends on work on emulation of specialization constants for other targets, so I'd wait to see how that ends and what we can do with that.